### PR TITLE
[Vis Augmenter] Add tests in OSD & AD plugin

### LINKED ISSUE #426: [Feature Anywhere] Add tests to create `augment-vis` saved objects for AD
Coming from https://github.com/opensearch-project/OpenSearch-Dashboards/issues/2880

Add flows from a Dashboard to link a detector to a visualization, and confirm the saved object is created as expected.

### LINKED ISSUE #428: [Feature Anywhere] Add tests for detector auto-creation from a visualization
Coming from https://github.com/opensearch-project/OpenSearch-Dashboards/issues/2880

Add a test flow where a detector config is auto-generated and created from a visualization, and that the saved object is created as expected.

### LINKED ISSUE #431: [Feature Anywhere] Add tests to show eligibility of the feature or not
Coming from https://github.com/opensearch-project/OpenSearch-Dashboards/issues/2880

The feature to augment a vis should only be visible when a vis is eligible (of type `line` and meeting other config requirements). Add tests to confirm the plugin options show or not based on eligibility.

### LINKED ISSUE #432: [Feature Anywhere] Add tests for new 'View events' flyout
Coming from https://github.com/opensearch-project/OpenSearch-Dashboards/issues/2880

New flyout should handle multiple scenarios and ensure it is rendered correctly:
- no events
- events for one plugin
- events for multiple plugins
- reaching plugin resource limit (if exists) and making scrollable or searchable, or showing most relevant plugin resources (?)

Last bullet point is an open question for now - will update once that decision has been finalized

### LINKED ISSUE #749: [BUG] re-add viz augmenter test
Had to revert to the viz augmenter test as they are consistently failing:

https://github.com/opensearch-project/opensearch-dashboards-functional-test/pull/748#issuecomment-1631668942

Tests seemed to have been ran when there was data existing in cluster. Also flakey setup code can be replaced with single import saved objects.

### PR DESCRIPTION
**Pull request by ohltyler**
### Description

Adds tests for the Vis Augmenter / Feature Anywhere feature. Includes new tests in core OSD + tests in AD plugin to test the integration of the plugin with the feature. Includes many helper fns in core OSD for plugins to use too bootstrap a test cluster with appropriate data to test plugin integration (e.g., create a dashboard with a feature-eligible visualization and sufficient sample data).

NOTE: there may be follow-ups to refactor when these new test suites are run, and some existing script cmds may need to be updated.

### Issues Resolved

Closes #426 
Closes #428 
Closes #431 
Closes #432 
Closes #749 

### Check List

- [x] Commits are signed per the DCO using --signoff

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by ohltyler** [comment]
@kavilla fixed NPE in latest commit. Regarding saved obj imports, I can explore doing that for some of the functionality. But a lot of the UI-based setup is through functions that take different parameters to be able to create different visualizations specific for a certain test suite or certain plugin testing the integration. I guess this could also be done with pre-created saved obj fixtures, but my concern is it could lead to a large number of json files for that.

I can write this up in a separate issue. For 2.9, I would prefer to get the current implementation merged to unblock. Thanks for taking the time to help debug as well :)

**Comment by kavilla** [comment]
@ohltyler I understand but with the NPE check I was not able get to actual test cases locally because it fails to succeed the setup steps. Im running the tests with cypress open from my Macbook and pointing to my EC2 instance.

Im willing to merge if not flakey. Do you have a github actions link with the cypress test passing consistently that we can feel confident in adding this?

**Comment by ohltyler** [comment]
@kavilla i've fixed the setup failures caused by a non-empty cluster putting pages into different state. Besides that, I've always been able to pass successfully on my local 3.0 cluster. Now it passes with existing data / saved objs on my local as well. Please try with the latest if you are still seeing issues

**Comment by kavilla** [comment]
> @kavilla i've fixed the setup failures caused by a non-empty cluster putting pages into different state. Besides that, I've always been able to pass successfully on my local 3.0 cluster. Now it passes with existing data / saved objs on my local as well. Please try with the latest if you are still seeing issues

> I'm still trying to figure out how to easily run on a non-3.0 branch since none of the tests are running since it fails to bring up 3.0 clusters to begin with...

@ohltyler I have created a draft PR and recreated the feature anywhere branch (since it was already merge). So any changes on the cypress feature/feature-anywhere branch should be pulled in here:
https://github.com/opensearch-project/OpenSearch-Dashboards/pull/4561

**Comment by ohltyler** [comment]
For merging to main, I'm seeing the following 2 issues:
1. When running the AD test workflows, all tests are failing due to elements not found. And when looking at the [workflow logs](https://github.com/opensearch-project/opensearch-dashboards-functional-test/actions/runs/5536830828/jobs/10105001646?pr=752), looks like AD Dashboards plugin is not being added into the cluster, hence all tests fail
2. When running in AD plugin, it is failing due to the following issue. It isn't seen on 2.x/2.9/etc., but I'm able to reproduce locally on a 3.0 build:
```
http://localhost:5601/api/console/proxy

We attempted to make an http request to this URL but the request failed without a response.

We received this error at the network level:

  > Error: connect ECONNREFUSED 127.0.0.1:5601
```
Workflow ref: [here](https://github.com/opensearch-project/anomaly-detection-dashboards-plugin/actions/runs/5524853110/jobs/10077722102)

**Comment by kavilla** [comment]
@ohltyler test seem to be passing well on the cypress test run for OpenSearch Dashboards. But can you check out the linter issues?

File: cypress/utils/dashboards/vis-augmenter/helpers.js [hunk 1]
```diff
@@ -0,0 +1,321 @@
...
+    apiRequest(
+      `api/saved_objects/index-pattern/${indexPatternName}`,
+      'DELETE'
+    ).then(() => {
+      apiRequest(
+        `
+api/opensearch-dashboards/management/saved_objects/_find?perPage=5000&page=1&fields=id&type=config&type=url&type=index-pattern&type=query&type=dashboard&type=visualization&type=visualization-visbuilder&type=augment-vis&type=search&sortField=type`,
+        'GET'
+      ).then((response) => {
+        response.body.saved_objects?.forEach((obj) => {
```
Comment (kavilla) [inline cypress/utils/dashboards/vis-augmenter/helpers.js:61]: Do you think we can do the following to pass the linter issues:

```js
if (!response.body.saved_objects) return;
response.body.saved_objects.forEach((obj) => {
```

Comment (ohltyler) [reply cypress/utils/dashboards/vis-augmenter/helpers.js]: done

**Review by kavilla — APPROVED** [review]

**Comment by kavilla** [comment]
Thanks for fixing! Yeah we should either update our eslint or add a parser to let us use optional chaining.

**Comment by ohltyler** [comment]
@SuZhou-Joe Can you help take a look here, since you helped with the original merge? The blocking issue that caused the revert has been resolved.

**Review by SuZhou-Joe — APPROVED** [review]

**Pull request merged**